ARCHITECTURE.md

```diff
@@ -1,113 +1,280 @@
+
+DI also allows us to circumvent the problem of cross-scene references to common dependencies, even though we still have to manage the lifecycle of MonoBehaviour-based dependencies by marking them with DontDestroyOnLoad and destroying them manually when appropriate.
+
+> __Note__:
+>
+> `ApplilcationController` inherits from the `VContainer`'s `LifetimeScope` - a class that serves as a dependency injection scope and bootstrapper, where we can bind dependencies. Scene-specific State classes inherit from `LifetimeScope` too.
+> 
+> In Inspector we can choose a parent scope for any `LifetimeScope`s - it’s useful to set a cross-scene reference to some parent scope (most commonly `ApplicationController`). This allows us to bind our scene-specific dependencies, and still have easy access to the global dependencies of the `ApplicationController` in our State-specific version of a `LifetimeScope` object.
+
+### Client/Server code separation
+
+Original problem: reading code with client and server code mixing together adds complexity and makes it easier to make mistakes. That code will often run in a single context, either client or server.
+
+Boss Room explored different client-server code separation approaches. For readers that have been following us since the beginning, we decided in the end to revert our initial client/server/shared assemblies to a more classic domain driven assembly architecture, but still keeping certain more complex classes with a client/server separation.
+
+Our initial thinking was separating assemblies by client and server would allow for easier porting to Dedicated Game Server (DGS) afterward. You’d only need to strip a single assembly to make sure that code only runs when necessary.
+
+Issues with this approach:
+
+- Callback hell: this makes code that should be trivial, too complex. You can look at our different action implementations in our 1.3.1 version to see this.
+- Lots of components could be single simple classes instead of 3 class horrors.
+
+We figured in the end this was not needed since:
+
+- You can ifdef out single classes, there’s no need for asmdef stripping.
+- ifdeffing classes isn’t 100% required. It’s a compile time insurance that certain parts of client side code will never run, but isn’t purely required.
+  - We realized the little pros that’d help with stripping whole assemblies out in one go were outweighed by the complexity this added to the project.
+- Most client/server class couples are tightly coupled and will call one another (they are two splitted implementations of the same logical object after all). Separating them in different assemblies makes you create “bridge classes” so you don’t have circular dependencies between your client and server assemblies. By putting your client and server classes in the same assemblies, you allow those circular dependencies in those tightly coupled classes and make sure you remove unnecessary bridging and abstractions.
```

Comment:
```suggestion
- Most client/server class couples are tightly coupled and will call one another; they are two split implementations of the same logical object. Separating them into different assemblies forces you to create “bridge classes” in order to avoid circular dependencies between your client and server assemblies. By putting your client and server classes in the same assemblies, you allow those circular dependencies in those tightly coupled classes and make sure you remove unnecessary bridging and abstractions.
```